### PR TITLE
[mlir][vector] Add vector.step operation

### PR DESCRIPTION
**Pull request by c-rhodes**
This patch adds a new vector.step operation to the Vector dialect. It produces a linear sequence of index values from 0 to N, where N is the number of elements in the result vector, and can be used to create vectors of indices.

It supports both fixed-width and scalable vectors. For fixed the canonical representation is `arith.constant dense<[0, .., N]>`. A scalable step cannot be represented as a constant and is lowered to the `llvm.experimental.stepvector` intrinsic [1].

This op enables scalable vectorization of linalg.index ops, see #96778. It can
also be used in the SparseVectorizer in-place of lower-level stepvector
intrinsic, see [2] (patch to follow).

[1] https://llvm.org/docs/LangRef.html#llvm-experimental-stepvector-intrinsic
[2] https://github.com/llvm/llvm-project/blob/acf675b63f9426e61aac2155e29280f7d21f9421/mlir/lib/Dialect/SparseTensor/Transforms/SparseVectorization.cpp#L385-L388

### CODE REVIEW

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td [hunk 1]
```diff
@@ -3017,6 +3017,35 @@ def Vector_ScanOp :
...
+//===----------------------------------------------------------------------===//
+
+def Vector_StepOp : Vector_Op<"step", [Pure]> {
+  let summary = "A linear sequence of values from 0 to N";
+  let description = [{
+    A `step` operation produces an index vector, i.e. a 1-D vector of values of
+    index type that represents a linear sequence from 0 to N, where N is the
+    number of elements in the `result` vector.
+
+    Supports fixed-width and scalable vectors. For fixed the canonical
```
Comment (banach-space) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:3031]:
```suggestion
    Supports fixed-width and scalable vectors. For a fixed-width step vector, the canonical
```

File: mlir/test/Dialect/Vector/ops.mlir [hunk 2]
```diff
@@ -1171,4 +1171,11 @@ func.func @from_elements(%a: f32, %b: f32) -> (vector<f32>, vector<1xf32>, vecto
   // CHECK: vector.from_elements %[[b]], %[[b]], %[[a]], %[[a]] : vector<2x2xf32>
   %3 = vector.from_elements %b, %b, %a, %a : vector<2x2xf32>
   return %0, %1, %2, %3 : vector<f32>, vector<1xf32>, vector<1x2xf32>, vector<2x2xf32>
-}
\ No newline at end of file
+}
+
+// CHECK-LABEL: @step
+func.func @step() {
+  %0 = vector.step : vector<2xindex>
```
Comment (banach-space) [inline mlir/test/Dialect/Vector/ops.mlir:1178]: Missing `// CHECK: vector.step`?

File: mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir [hunk 3]
```diff
@@ -2621,3 +2621,14 @@ func.func @vector_from_elements_0d(%a: f32) -> vector<f32> {
...
   return %0 : vector<f32>
 }
+
+// -----
+
+// CHECK-LABEL: @vector_step
+// CHECK: %[[STEPVECTOR:.*]] = llvm.intr.experimental.stepvector : vector<[4]xi64>
+// CHECK: %[[CAST:.*]] = builtin.unrealized_conversion_cast %[[STEPVECTOR]] : vector<[4]xi64> to vector<[4]xindex>
+// CHECK: return %[[CAST]] : vector<[4]xindex>
+func.func @vector_step() -> vector<[4]xindex> {
```
Comment (banach-space) [inline mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir:2631]:
```suggestion
func.func @vector_step_scalable() -> vector<[4]xindex> {
```

Let's start properly labeling tests which are intended for scalable vectors only.

**Review by banach-space — APPROVED** [review]
Makes sense, thanks! I've a few nits, but nothing major (LGTM modulo the comments)

What's the intended use for this? Worth adding a note in the summary.

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td [hunk 4]
```diff
@@ -3017,6 +3017,35 @@ def Vector_ScanOp :
...
+
+def Vector_StepOp : Vector_Op<"step", [Pure]> {
+  let summary = "A linear sequence of values from 0 to N";
+  let description = [{
+    A `step` operation produces an index vector, i.e. a 1-D vector of values of
+    index type that represents a linear sequence from 0 to N, where N is the
+    number of elements in the `result` vector.
+
+    Supports fixed-width and scalable vectors. For a fixed-width `step` vector,
+    the canonical representation is `arith.constant dense<[0, .., N]>`. A
```
[the remark this answers is not in the document]
Comment (c-rhodes) [reply mlir/include/mlir/Dialect/Vector/IR/VectorOps.td]: N-1, fixed thanks!

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td [hunk 5]
```diff
@@ -3017,6 +3017,35 @@ def Vector_ScanOp :
...
+  let description = [{
+    A `step` operation produces an index vector, i.e. a 1-D vector of values of
+    index type that represents a linear sequence from 0 to N-1, where N is the
+    number of elements in the `result` vector.
+
+    Supports fixed-width and scalable vectors. For a fixed-width `step` vector,
+    the canonical representation is `arith.constant dense<[0, .., N-1]>`. A
+    scalable step cannot be represented as a constant and is lowered to the
+    [llvm.experimental.stepvector](https://llvm.org/docs/LangRef.html#llvm-experimental-stepvector-intrinsic)
+    intrinsic.
```
Comment (MacDue) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:3035]: I don't think a Vector level operation has a specific lowering (not all paths out of Vector go through LLVM). You could maybe reword it slightly to say "When lowering to LLVM a scalable step can be...". 

However, I'm not sure it's necessary to talk about how an op is lowered (i.e. specific intrinsic) within the operation description.

**Review by MacDue — APPROVED** [review]
Generally LGTM, just a note on the docs:

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp [hunk 6]
```diff
@@ -6316,6 +6316,20 @@ OpFoldResult SplatOp::fold(FoldAdaptor adaptor) {
...
+//===----------------------------------------------------------------------===//
+
+OpFoldResult StepOp::fold(FoldAdaptor adaptor) {
+  auto resultType = cast<VectorType>(getType());
+  if (resultType.isScalable())
+    return nullptr;
+  SmallVector<APInt> indices;
+  for (unsigned i = 0; i < resultType.getNumElements(); i++)
+    indices.push_back(APInt(/*width=*/64, i));
+  return DenseElementsAttr::get(resultType, indices);
```
Comment (dcaballe) [inline mlir/lib/Dialect/Vector/IR/VectorOps.cpp:6330]: Why do we want to canonicalize the fixed-length variant in the Vector dialect to something else that is not a `vector.step`? I'd suggest that we use `vector.step` as canonical representation for both fixed and scalable versions for consistency. Is there any reason to not doing so?

Comment (c-rhodes) [reply mlir/lib/Dialect/Vector/IR/VectorOps.cpp:6330]: Purely for convenience to be honest, this seemed like the simplest way to add the op. I know the downside is information is lost when folding to a constant (i.e. it's a constant step of one), as well as consistency as you point out, so am happy to iterate on this if folks would like to see this done differently.

Comment (dcaballe) [reply mlir/lib/Dialect/Vector/IR/VectorOps.cpp:6330]: Thanks for clarifying! I think consistency is really important to make sure fixed and scalable worlds are aligned when possible and we don't end up with a different set of patterns and problems when we shouldn't. `vector.step` looks like a good abstraction for both fixed and scalable vectors so I'm +1 to removing this canonicalization pattern and let `vector.step` be the canonical representation also for fixed vectors.
